spec/index.bs

```diff
@@ -1033,17 +1033,48 @@ its mapping.
      mapping.
 
 <script type=idl>
+typedef [EnforceRange] unsigned long GPUMapUsageFlags;
+interface GPUMapUsage {
+    const GPUMapUsageFlags READ  = 0x0001;
+    const GPUMapUsageFlags WRITE = 0x0002;
+    const GPUMapUsageFlags READ_WRITE = 0x0003;
+};
+
+interface GPUMappedSubrange {
+    ArrayBuffer mapped;
+};
+
 [Serializable]
 interface GPUBuffer {
-    Promise<ArrayBuffer> mapReadAsync();
-    Promise<ArrayBuffer> mapWriteAsync();
-    void unmap();
+    Promise<void> mapRangeAsync(GPUBufferMapFlags flags, unsigned long long offset = 0, unsigned long long size = 0);
+    GPUMappedSubrange? getMappedSubrange(unsigned long long offset = 0, unsigned long long size = 0);
+    void unmapRange(unsigned long long offset = 0, unsigned long long size = 0);
 
     void destroy();
 };
 GPUBuffer includes GPUObjectBase;
 </script>
 
+If size is zero, {{GPUBuffer/map}} treats it as an effective size of buffer-length-starting-at-offset.
+Effective size of a mapping must not be zero.
+{{GPUBuffer/map}} fails (and rejects) if the range overlaps with already-mapped ranges in GPUBuffer.
+{{GPUBuffer/map}}'s Promise resolves before {{GPUFence/onCompletion}}'s Promise for any
```

Comment:
Let's consider this scenario:
  - user submits work X on part A of a buffer
  - user requests to map part B of the buffer, B doesn't intersect with A
  - user submits work Y on part A of the buffer

The rules you specify basically guarantee that the mapping resolves while part A is still being used by the GPU. So this proposal requires the implementation to track buffer sub-ranges independently. I think it's essential to write it down clearly (i.e. what the implementations are supposed to do) instead of implying.